apps/sms/views/shared/js/navigation.js

```diff
@@ -1,361 +1,726 @@
+  }
+
+  return findViewFromName(hash);
+}
+
+/**
+ * Tries to find the correct view from the location.
+ * First it tries to find the view using the location's hash; if
+ * findViewFromHash returns something, then we check whether the found view is
+ * indeed part of this document usinf the full URL.
+ * If findViewFromHash returns nothing, we try to find a view from the full URL.
+ *
+ * @param {Location} location The content of window.location.
+ * @returns {Object} The found view, or null.
+ */
+function findViewFromLocation(location) {
+  var pathName = location.pathname;
+  if (pathName.endsWith('/')) {
+    pathName += 'index.html';
+  }
+
+  var viewFromHash = findViewFromHash(location.hash);
+  var viewFromPath = VIEWS.find(
+    (object) => pathName.endsWith('/' + object.url)
+  ) || null;
+
+  if (viewFromHash && viewFromHash.partOf &&
+      viewFromHash.partOf !== viewFromPath.name) {
+    console.error(
+      'The view', viewFromHash.name,
+      'found from the hash', location.hash,
+      'is not part of view', viewFromPath.name
+    );
+    viewFromHash = null;
+  }
+
+  return viewFromHash || viewFromPath;
+}
+
+/**
+ * Finds a view from its name.
+ *
+ * @param {String} name Name of the view to find.
+ * @returns {Object} The found view, or null.
+ */
+function findViewFromName(name) {
+  return VIEWS.find((object) => {
+    return object.name === name;
+  }) || null;
+}
+
+/**
+ * Sets up the state to handle the navigation.
+ * Generally we're waiting for any ongoing transition to end before starting a
+ * new one. However if the ongoing transition was started with opts.partial ==
+ * true, we merely augment the ongoing transition with the new information,
+ * especially arguments.
+ *
+ * @param {Object} opts options
+ * @param {String} opts.newPanel The panel we'll switch to.
+ * @param {Object} opts.args The arguments we'll use for the next panel.
+ * @param {Boolean} [opts.partial] If present and true, this new navigation
+ * request is a partial one missing information (eg arguments) and we'll call
+ * startNavigation once again without this property once we'll know the missing
+ * bits.
+ *
+ * @returns {Promise} Resolved when we could start the navigation (when an
+ * previous ongoing one is finished).
+ */
+function startNavigation(opts) {
+  var { newPanel, args, partial } = opts;
+
+  debug('startNavigation(', JSON.stringify(opts), ')');
+  if (navigationTransition && !navigationTransition.partial) {
+    debug('startNavigation: previous navigation exists, defering.');
+    return navigationTransition.defer.promise.then(
+      () => startNavigation(opts)
+    );
+  }
+
+  if (!navigationTransition) {
+    var defer = Utils.Promise.defer();
+
+    navigationTransition = {
+      stepsRun: [],
+      defer
+    };
+  }
+
+  navigationTransition.partial = partial;
+
+  if (args && ('meta' in args)) {
+    throw new Error('Arguments cannot have a `meta` property.');
+  }
+
+  navigationTransition.transitionArgs = {
```

Comment:
also I agree that we have the current args both in transitionArgs.meta.next.args and in the "root" of transitionArgs; the initial idea was that we want next.args because we had prev.args. And it's also a lot easier to have them at the 1st level of properties.